[FLINK-24657][runtime] Added metric of the total real size of input/o…

akalash opened the pull request:
…utput buffers queue

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*This PR adds two new metrics - outputQueueSize and inputQueueSize*


## Brief change log

  - *PipelinedSubpartitin#totalNumberOfBytes is calculated only for data buffer*
  - *Output queue buffers size is calculated based on numBytesOut metric and totalNumberOfBytes*


## Verifying this change

This change added tests and can be verified as follows:

  - *Added new test for checking queue size for PartitionResult*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java:
```diff
@@ -111,6 +111,17 @@ public int getNumberOfQueuedBuffers() {
+        long totalNumberOfBytes = 0;
+
+        for (ResultSubpartition subpartition : subpartitions) {
+            totalNumberOfBytes += Math.max(0, subpartition.getTotalNumberOfBytes());
```
I just notice that the access to getTotalNumberOfBytes is unsynchronized here.

pnowojski replied:
Rename `getSizeOfQueuedBuffers` to `getSizeOfQueuedBuffersUnsafe`?

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -497,6 +497,12 @@ public int unsynchronizedGetNumberOfQueuedBuffers() {
+    @Override
+    public long unsynchronizedGetSizeOfQueuedBuffers() {
+        // Pretty rough approximation of real queue size.
+        return Math.max(unsynchronizedGetNumberOfQueuedBuffers() * bufferSize, 0);
```
This is pretty bad since the queue can hold events that have different sizes, buffers of old buffer size, and unfinished buffers. But it is the easiest way to calculate queue size without dramatically rewriting code.

pnowojski replied:
I have a feeling that in this implementation this metric doesn't really tell us anything more compared to output queue length and desired buffer size. Probably we should provide a more accurate estimate/metric or just drop it.

The question would be what to do with this for the input bytes length. As this is only for local input channels, maybe that's a good enough estimation?

akalash replied:
I think for only local channels is a good trade off

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -535,7 +541,9 @@ private void updateStatistics(BufferConsumer buffer) {
     private void updateStatistics(Buffer buffer) {
-        totalNumberOfBytes += buffer.getSize();
+        if (buffer.isBuffer()) {
+            totalNumberOfBytes += buffer.getSize();
```
NB! Right now, The totalNumberOfBytes calculate only the size of the data buffers(without event) which broke the old behavior but as I can see we use it only in toString. So maybe it is not a problem but if it is, we can add another variable for data buffers size calculation.

pnowojski replied:
But why are you changing this in the first place? What's your motivation?

pnowojski replied:
scratch that. Now I can see why have you change this. I think this change would be fine, assuming we want the `outputQueueSize` to ignore events. If `outputQueueLength` is ignoring events, then I would vote for `outputQueueSize` to ignore them as well. In other words, they should stay in sync.

akalash replied:
`outputQueueLength` takes into account all buffers including the event. So I implemented the same logic for `outputQueueSize`

akalash reviewed (commented):
The current implementation uses the existing numBytesOut metric and ResultSubpartition#totalNumbersOfBytes variable for calculating output buffers queue size. But it has a problem with the calculation of input buffers queue size for the local channel since we don't have information about the number of bytes that are in the queue. 

Other solutions:

- Calculation numBytesOut for each subpartition separately which requires having a map inside ResultPartition or call the ResultSubpartition every time when we want to calculate the size. In both ways, the solution looks expensive.
- Iterate over queue and calculate the size of the total buffer. This solution is most accurate but it is pretty expensive to calculate this metric.

zentol commented:
Remember to update the documentation as well: https://nightlies.apache.org/flink/flink-docs-master/docs/ops/metrics/#system-metrics

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -433,6 +433,24 @@ public int getNumberOfQueuedBuffers() {
+                }
+
+                return totalSize;
+            } catch (Exception ignored) {
```
log exception with debug?

pnowojski reviewed (commented):
With this proposal, I would be leaning toward dropping the output bytes length metric, and just keeping input one.

However what about other ideas, especially for the output metric, like having two `long` byte counters: `totalBytesWritten` and `totalBytesSent` and then deducing 
```
long getSizeOfQueuedBuffers() {
  return totalBytesWritten - totalBytesSent;
}
```
?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java:
```diff
@@ -111,6 +111,17 @@ public int getNumberOfQueuedBuffers() {
+            totalNumberOfBytes += Math.max(0, subpartition.getTotalNumberOfBytes());
+        }
+
+        return numBytesOut.getCount() - totalNumberOfBytes;
```
This metric is updated only after buffer writers are finished, which will lead to incorrect results here (negative numbers). I would suggest to calculate and update the `numBytesOut` every record. But as this would be now done on the per record code path, I would suggest to refactor the code to have `numBytesOut` as a simple `private long` field to avoid potential performance issues.

akalash replied:
I agree. I implemented the new field `totalWrittenBytes` which will be updated on every record.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java:
```diff
@@ -64,11 +64,13 @@
     // task level output metrics: Shuffle.Netty.Output.*
 
     private static final String METRIC_OUTPUT_QUEUE_LENGTH = "outputQueueLength";
+    private static final String METRIC_OUTPUT_QUEUE_SIZE = "outputQueueSize";
```
+1 for picking a good name

pnowojski reviewed (commented):
Before merging, we should also update the documentation. And in the docs I would mention that local input channels are only estimated.

akalash commented:
I have updated the PR with described changes. I added the new commit but if it would be more convenient to you I can squash it with the previous one.
I also added the documentation.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java:
```diff
@@ -64,6 +64,8 @@
 
     private TimerGauge backPressuredTimeMsPerSecond = new TimerGauge();
 
+    private long totalWrittenBytes;
```
can we use this variable to replace `numBytesOut` `Counter` and metric?

akalash replied:
Of course, we can rewrite the metric in such a way that instead of passing the counter it can call the method for receiving the number of bytes. But it definitely is not one line change and if we do so unfortunately semantic for `numBuffersOut` and `numBytesOut` will be different which is not tragic but anyway. 
So we can do this but I don't sure that it makes sense to do it here. Or do you think that the separate commit under this ticket is ok?

pnowojski replied:
If we decide to do that, I think it should be done under this ticket to avoid acquiring a technological debt (in a separate commit).

However keep in mind that I haven't thought deeply about this deduplication. If it adds a lot of complexity it might be actually cleaner to have them duplicated.

akalash replied:
I have thought about that. I think it is better to leave them duplicated since right now we have `TaskIOMetricGroup` which handles the metrics the same - `Counter` + `MeterView` but if we want to change it we need to do at least the following:

- We need to adapt `MeterView` to receive value directly from some entity rather than from `Counter`(or create a wrapper for counter). This is not a big problem but anyway, this should be done.
- We need to initialize one particular metric differently since we won't be able to create this metric in `TaskIOMetricGroup` constructor. This is a more serious problem because it can lead to confusion if we start to initialize metrics from one group differently in different places.
- We need to implement the same logic of `totalWrittenBytes` in other subclasess of `ResultPartition` since right now it is just a helper for calculation of metric in one specific implementation but if we decide to make it as metric we need to have it everywhere(at least in `SortMergeResultPartition`). Again, it is not a big problem but just an extra complication.

So right now I think that it doesn't make sense to change anything in `TaskIOMetricGroup` but in general, it is a good idea to think about our metrics. Because the metric which collects data directly from the entity by demand looks more flexible than the metric that passes the counter(or other metric-related things) to the entity. But this is not today's discussion.

pnowojski reviewed (commented):
Thanks for the updates. After an offline discussion +1 for this idea. Indeed I've missed that this metric will be inaccurate just for `LocalInputChannels`, which is probably good enough.

pnowojski reviewed (approved):
LGTM % squashing of the 2nd commit

pnowojski commented on docs/content.zh/docs/ops/metrics.md:
```diff
@@ -916,12 +916,17 @@ Metrics related to data exchange between task executors using netty network comm
     </tr>
+    <tr>
+      <td>inputQueueSize</td>
+      <td>The real size of queued input buffers in bytes. The size for local input channels is only estimated.</td>
```
On a second thought. Shouldn't we actually ignore local input channels? 

1. It would be kind of consistent with the other metrics
2. Otherwise, don't we double account those bytes on the output and on the input?

akalash replied:
In fact, you are right we count bytes twice for the local input channel but we do the same thing for `outputQueueLength`/`inputQueueLength`. But maybe it is not a bad idea to always return zero for local input channels since they take the data directly from output.

akalash commented:
@pnowojski, I have updated the PR with removed queue size for the local input channel. I also have updated the documentation accordingly.
